From: NeilBrown <neilb@suse.de>
Git-commit: 1fdd6fc92f330b81604c9a4f892f713775a9751a
Date: Wed, 14 Mar 2012 18:04:56 +1100
Subject: [PATCH] md: teach sync_page_io about new_data_offset.
Patch-mainline: maybe 3.5
References: 

Some code in raid1 and raid10 use sync_page_io to
read/write pages when responding to read errors.
As we will shortly support changing data_offset for
raid10, this function must understand new_data_offset.

So add that understanding.

Signed-off-by: NeilBrown <neilb@suse.de>

---
 drivers/md/md.c |    4 ++++
 1 file changed, 4 insertions(+)

--- linux-3.0-SLE11-SP3.orig/drivers/md/md.c
+++ linux-3.0-SLE11-SP3/drivers/md/md.c
@@ -861,6 +861,10 @@ int sync_page_io(mdk_rdev_t *rdev, secto
 		rdev->meta_bdev : rdev->bdev;
 	if (metadata_op)
 		bio->bi_sector = sector + rdev->sb_start;
+	else if (rdev->mddev->reshape_position != MaxSector &&
+		 (rdev->mddev->reshape_backwards ==
+		  (sector >= rdev->mddev->reshape_position)))
+		bio->bi_sector = sector + rdev->new_data_offset;
 	else
 		bio->bi_sector = sector + rdev->data_offset;
 	bio_add_page(bio, page, size, 0);
